File: interpreter/interpreter/lexer.py
from .token import TokenType, Token

class LexerException(Exception):
    pass

class Lexer():

    def __init__(self) -> None:
        self._pos: int = -1
        self._current_char: str = ''
        self._text: str = ''

    def next(self) -> Token:
        while self._current_char != None:
            if self._current_char == ' ':
                self._skip()
                continue

            char = self._current_char
         
            if self._current_char.isdigit():
                number = self._float()
                if '.' in number:
                    return Token(TokenType.FLOAT, number)
                else:
                    return Token(TokenType.INTEGER, number)
            elif self._current_char == '+':
                self._forward()
                return Token(TokenType.PLUS, char)
            elif self._current_char == '-':
                self._forward()
                return Token(TokenType.MINUS, char)
            elif self._current_char == '*':
                self._forward()
                return Token(TokenType.MUL, char)
            elif self._current_char == '/':
                self._forward()
                return Token(TokenType.DIV, char)
            elif self._current_char == '^':
                self._forward()
                return Token(TokenType.POW, char)
            elif self._current_char == '(':
                char = self._current_char
                self._forward()
                return Token(TokenType.LPAREN, char)
            elif self._current_char == ')':
                char = self._current_char
                self._forward()
                return Token(TokenType.RPAREN, char)
            raise LexerException(f"Undefined token: {char}")
        return Token(TokenType.EOS, None)

    def _forward(self):
        self._pos += 1
        if self._pos >= len(self._text):
            self._current_char = None
        else:
            self._current_char = self._text[self._pos]
        
    def _skip(self):
        while self._current_char == ' ':
            self._forward()

    def _integer(self) -> str:
        result: list = []
        while self._current_char and self._current_char.isdigit():
            result.append(self._current_char)
            self._forward()
        return ''.join(result)
    
    def _float(self) -> str:
        lhs_part = self._integer()
        if self._current_char == '.':
            self._forward()
            rhs_part = self._integer()
            return f"{lhs_part}.{rhs_part}"
        return lhs_part
    
    def init(self, text: str):
        self._text = text
        self._pos = -1
        self._forward()

File: interpreter_rust/interpreter/src/lexer.rs
use super::token::{Token, TokenType};

#[derive(Debug)]
pub(crate) struct Lexer {
    position: usize,
    text: Vec<char>
}

impl Lexer {
    pub fn from<'a>(text: &'a str) -> Self {
        Self {
            position: 0,
            text: text.chars().collect()
        }
    }

    pub fn next(&mut self) -> Token {
        while let Some(current_char) = self.text.get(self.position) {
            if current_char.is_whitespace() {
                self.skip();
                continue;
            }

            if current_char.is_ascii_digit() {
                return self.number();
            } else {
                self.position += 1;

                let token_type = match current_char {
                    '+' => TokenType::Add,
                    '-' => TokenType::Sub,
                    '*' => TokenType::Mul,
                    '/' => TokenType::Div,
                    '^' => TokenType::Pow,
                    '(' => TokenType::Lp,
                    ')' => TokenType::Rp,
                    _ => panic!("Undefined token {:?}", current_char)
                };

                return Token::new(token_type, Some(current_char.to_string()));
            }
        }
        return Token::new(TokenType::Eos, None);
    }

    fn skip(&mut self) {
        while let Some(current_char) = self.text.get(self.position) {
            if !current_char.is_whitespace() {
                break;
            }
            self.position += 1;
        }
    }

    fn number(&mut self) -> Token {
        let mut integer_part = self.integer();

        if let Some(current_char) = self.text.get(self.position) {
            if current_char == &'.' {
                integer_part.push(*current_char);
                self.position += 1;
                let float_part = self.integer();

                return Token::new(TokenType::Float, Some(integer_part + &float_part));
            }
        }

        Token::new(TokenType::Integer, Some(integer_part))
    }

    fn integer(&mut self) -> String {
        let mut number = Vec::<char>::new();

        while let Some(current_char) = self.text.get(self.position) {
            // for now we use only decimal numbers
            if !current_char.is_ascii_digit() {
                break;
            }

            number.push(*current_char);
            self.position += 1;
        }

        number.into_iter().collect::<String>()
    }
}


#[cfg(test)]
mod tests {
    use super::{Lexer, Token, TokenType};

    #[test]
    fn create_lexer_from() {
        let mut lexer = Lexer::from("2 + 5");
        assert_eq!("Lexer { position: 0, text: ['2', ' ', '+', ' ', '5'] }", format!("{:?}", lexer));

        lexer = Lexer::from(&String::from("2 + 5"));
        assert_eq!("Lexer { position: 0, text: ['2', ' ', '+', ' ', '5'] }", format!("{:?}", lexer));
    }

    #[test]
    fn tokenize() {
        let mut lexer = Lexer::from("4 - 5");

        let lhs = Token::new(TokenType::Integer, Some("4".to_string()));
        let mut token = lexer.next();
        assert_eq!(lhs, token);

        let op = Token::new(TokenType::Sub, Some("-".to_string()));
        token = lexer.next();
        assert_eq!(op, token);

        let rhs = Token::new(TokenType::Integer, Some("5".to_string()));
        token = lexer.next();
        assert_eq!(rhs, token);
    }

    #[test]
    fn tokenize_empty_str() {
        let mut lexer = Lexer::from("");
        let empty_token = Token::new(TokenType::Eos, None);
        
        assert_eq!(empty_token, lexer.next());
    }
}